climbing-stairs/bhyun-kim.py

```diff
@@ -0,0 +1,39 @@
+"""
+70. Climbing Stairs
+https://leetcode.com/problems/climbing-stairs/description/
+
+Solution:
+    This is a combinatorial problem that allows duplicates.
+    We can use the formula for combinations to solve this problem.
+    Let's suppose that n is the number of steps, and k is the number of 2-steps.
+    Then, the number of ways to climb the stairs is n!/((n-2k)!k!).
+    We can iterate through all possible values of k, and calculate the number of ways to climb the stairs.
+
+    1. Create a dictionary to store the factorials of numbers from 0 to n.
+    2. Calculate the factorials of numbers from 0 to n.
+    3. Iterate through all possible values of k from 0 to n//2.
+
+
+Time complexity: O(n)
+Space complexity: O(n)   
+"""
+
+from collections import defaultdict
+
+
+class Solution:
+    def climbStairs(self, n: int) -> int:
+        factorials = defaultdict(int)
+        factorials[0] = 1
+
+        fact = 1
+        for i in range(1, n + 1):
+            fact = fact * i
+            factorials[i] = fact
+
+        output = 0
+
+        for i in range(n // 2 + 1):
```

Comment:
파이썬이 2로 나눗셈 할 때 최적화를 해주는 지는 모르겠으나, 약간의 속도 팁을 내자면 shift 연산으로 2를 나눈 거랑 동일한 결과를 낼 수 있습니다.
shift 연산이 나눗셈보다 월등히 빠르기 때문에 이런 점 고려해보시면 좋을 것 같아요 ㅎㅎ

```suggestion
        for i in range((n >> 1) + 1):
```

Reply:
shift 연산이 나눗셈보다 월등히 빠르다는 근거가 무엇인지 궁금합니다! (그냥 호기심입니당 😆)

Reply:
> @WhiteHyun 쉬프트 연산이 나눗셈보다 월등히 빠르다는 근거가 무엇인지 궁금합니다! (그냥 호기심입니당 😆)

[스택오버플로우에서 찾아봤는데요 이런 글이 있네요. ](https://stackoverflow.com/questions/37053379/times-two-faster-than-bit-shift-for-python-3-x-integers#:~:text=Integer%20division,small%20ints%20too.
) Python에서 나눗셈을 할 때, long_divrem()이라는 함수를 이용해서 몫과 나머지를 모두 구한 후 몫만 반환하기 때문이라는 설명이네요.